feat: enhance pathfinding options with maxCost, maxNodes, and allowedCallback

123FLO321 opened the pull request:
This pull request adds new constraints and customization options to the `Graph` class's `path()` method, enhancing its flexibility for pathfinding. It introduces support for limiting the maximum cost and maximum number of nodes in a path, as well as a callback to control which edges are allowed during path expansion. 
Tests for new cases are added. 

closes #46 

Also includes types (from @types/node-dijkstra) with modifications untill #45 is merged.

123FLO321 committed:
feat: add native types

123FLO321 committed:
feat: enhance pathfinding options with maxCost, maxNodes, and allowedCallback

(closes #46)

albertorestifo commented on libs/Graph.js:
```diff
@@ -192,14 +196,28 @@ class Graph {
+    const maxCost = hasMaxCost ? Number(options.maxCost) : undefined;
+    const hasMaxNodes = typeof options.maxNodes === 'number' && !Number.isNaN(options.maxNodes);
+    const maxNodes = hasMaxNodes ? Math.max(1, Math.floor(options.maxNodes)) : undefined;
+    const allowedCallback = typeof options.allowedCallback === 'function' ? options.allowedCallback : null;
```
What do you think of `canVisit` as a name for this argument?

albertorestifo commented on test/Graph.test.js:
```diff
@@ -327,4 +327,130 @@ describe("Graph", () => {
+      res.cost.must.equal(0);
+    });
+
+    it("receives correct argument shape (from, to, cost, accumulatedCost, depth)", () => {
```
If you have time, it would be good to add a test to verify the `depth` is calculated correctly:

A graph with different branches, should result in the right depth for each node.

albertorestifo reviewed (commented):
Thanks for your contribution! :)

Left a couple of notes, let me know what you think.

albertorestifo commented:
Ah also I see there are some errors from Prettier, can you please run it on the code? Thanks!

123FLO321 committed:
refactor: fix lint errors

123FLO321 committed:
refactor: rename allowed callback to can visit

123FLO321 committed:
build: separate test lint from test

123FLO321 committed:
refactor: fix lint errors

123FLO321 committed:
test: add tests for expected depths

123FLO321 commented:
@albertorestifo I fixed the lint issues (and also moved lint from the test script and explicitly run it in the workflow).
And also implemented the requested test.

123FLO321 committed:
test: add tests for expected weights

123FLO321 committed:
refactor: fix types for can visit

123FLO321 commented on libs/Graph.js:
```diff
@@ -192,14 +196,32 @@ class Graph {
+
     // Add the starting point to the frontier, it will be the first node visited
     frontier.set(start, 0);
+    depth.set(start, 1);
```
Should depth start at 0 or 1 for the root node?
